### ISSUE TITLE
build instructions

### ISSUE THREAD

**Issue by schriftgestalt** [Microsoft/Font-Validator#6]
The readme file mentions Linux and MacOS several times. Is it actually possible to build for this platforms? If so, how? I can build a windows .exe (after I installed mono. A link to the mono installer should be somewhere in the readme).

**Comment by anthrotype** [comment]
after you install mono on OS X, you should be able to simply run

    make

I don't have a mac at hand right now so can't try.

**Comment by schriftgestalt** [comment]
I did run make. But it produces Windows .exe files.

**Comment by anthrotype** [comment]
they are actually mono .exe files.
you run them by doing

    mono FontValidator.exe

**Comment by anthrotype** [comment]
http://www.mono-project.com/docs/about-mono/supported-platforms/osx/#using-mono-on-macos-x

**Comment by schriftgestalt** [comment]
Is that mentioned anywhere?

**Comment by anthrotype** [comment]
if not, it should ;)

**Comment by schriftgestalt** [comment]
I just googled this a bit and there are options to build regular MacOS apps from mono exe files. I didn't get it to work but someone with a bit knowledge of the Mono should be able to figure this out:

http://www.mono-project.com/archived/guiderunning_mono_applications/
http://www.mono-project.com/docs/tools+libraries/libraries/monomac/

**Comment by anthrotype** [comment]
yeah, that would be nice!
:+1:

**Comment by HinTak** [comment]
dotnet executables are actually platform-neutral; - think of the *.exe as java *.jar files. There is a compiler which turns text files into byte-codes, and a runtime to run the byte-codes.

I have in the past use mono's mkbundle to create native linux ELFs (and the same process does MACH-O for Mac OS X). There is no clear advantage for that, because you basically bundle 1/3 of the runtime with your application in the end, so your executable is just larger. It is not even faster.

**Comment by schriftgestalt** [comment]
> so your executable is just larger. It is not even faster.

But you can run it from finder. That is not important for the command line version, but for the GUI version it would be much better. And you can distribute binaries without the need to install the runtime. I read somewhere something that there should be a method that only includes the needed to classes to keep the size down.

**Comment by HinTak** [comment]
Yes and no. As I said, I have done that before for linux, and it is exactly the same process for Mac OS X. The generated platform-specific executable is not only larger, but also still have a runtime dependency on mono's runtime libraries for the byte-code interpreter. i.e all the *.dll's is gone, but you still needs libmono.dylib (for mac) or libmono.so (for linux).

In my previous project, going further - statically linking libmono into my platform-specific executable - to remove all dependency on the runtime library - is not a viable option due to licensing reasons, so I stopped there, after seeing my platform-neutral executable going from 5MB to 50MB and still depends on the runtime library.

There is also a somewhat general issue that multi-threadded applications tend to suffer performance-wise when statically linked. My previous project wasn't multi-threaded much, but Font Validator is quite heavily so.

The licensing issue (not to statically link non-open code with open code) is not the case here - and to be honest, I can do it and I haven't got round to, just because it isn't a priority, and quite far from it. I have enough to do as is.

Would you and/or you know somebody like to commission native MACH-O executable done?

**Comment by justanotherfoundry** [comment]
Sorry if this is a stupid question but do I have to install "mono" first?

**Comment by HinTak** [comment]
@justanotherfoundry : yes. The situation is entirely the same as java. If you develop, you need both the compiler and the runtime, if you are interested in running only, you still need the runtime. For Linux and windows, Ximian (or Linux distros) don't distinguish the two; but for Mac OS X, the last time I checked, they do distiguish that - you can download a runtime-only package, or a larger SDK package (which includes the runtime).

I feel this info should be a FAQ. I'll go and stick this on my wiki now.

**Comment by justanotherfoundry** [comment]
Ah, thanks. Got it to work.

**Comment by HinTak** [comment]
The native executable for Mac OS X is under 30MB; but I haven't managed to get the GUI to work ATM, and you'd still need to have freetype installed.

How happy are Mac OS X users on the command line? i.e. if I can't get the Font Val GUI fully working independent of a  mono installation?

**Comment by schriftgestalt** [comment]
I’m just used to run FontValildator with UI. But all other tools (ttx) are fine in the command line...

**Comment by HinTak** [comment]
Making the command-line part of Font Validator independent of a mono installation is fairly straight forward - I only had two issues to work around, one of which I anticipated before I started looking; I don't have an ETA for making the GUI fully working at the moment. I'll make some additional mono-independent Mac OS X binaries, the next time I build and upload a set of binaries.

**Comment by HinTak** [comment]
Okay, no more messing with command lines or environment variables. New mac os "FontVal-*.dmg" disk image:

http://sourceforge.net/projects/hp-pxl-jetready/files/Microsoft Font Validator/

You just download, double click to open the disk image, which contains exactly one icon to click. You double-click that, it launches the GUI !

I feel that I should get some donation on this, so please do make one (
http://sourceforge.net/p/hp-pxl-jetready/donate/) if you find this neat .

**Comment by HinTak** [comment]
Building with MS Visual Studio has been mostly re-added. 

Note that the Compat module (the module I wrote for replacing everything that Microsoft did not open) does not build with MS VC# yet. It should be straight-forward, it isn't done yet mainly because I don't use MS VC# regularly enough, nor windows itself for that matter :). Tips to go about that is here:
https://github.com/HinTak/Font-Validator/issues/8#issuecomment-186716648
At some point I'll have a go at that.

**Comment by twardoch** [comment]
Xamarin (recently acquired by Microsoft) promises to be able to run C# code possibly with a native Mac UI. The Xamarin.Mac package is said to be a "superset of MonoMac". Would that possibly be an option? 

https://developer.xamarin.com/guides/mac/getting_started/hello,_mac/
https://www.xamarin.com/download
https://developer.xamarin.com/api/root/monomac-lib/
http://www.mono-project.com/docs/tools+libraries/libraries/monomac/

**Comment by HinTak** [comment]
Thanks for the notes on Xamarin.Mac. The short answer is no, for the purpose of this bug report; OTOH, I have already been looking at Xamarin.Mac because it provides sysdrawing-coregraphics
( https://github.com/HinTak/Font-Validator/issues/7#issuecomment-187298106 ) which I may like to have to work around some other issues over there.

**Comment by HinTak** [comment]
Correction: Xamarin.Mac does not provide sysdrawing-coregraphics, but building sysdrawing-coregraphics from source depends on and requires Xamarin.Mac  .

BTW, Xamarin.Mac  is not a runtime environment  i.e. it does not magically let you run platform-neutral binary better - it is a set of dev tools and extra libraries which provides extra and deeper APIs into the Mac platform. You still need to write new Mac-specific new code to use it.

**Comment by twardoch** [comment]
Supposedly Xamarin will be free for VS users: http://thenextweb.com/voice/2016/03/30/tnw-logos-designed-audience/

**Comment by HinTak** [comment]
It is good news, indeed, in that people with C# knowledge tend to be from the VS studio camp. I am in the minority in that my c# knowledge is almost all with mono...

BTW, Xamarin is basically the commercial entity behind the mono project.
